pyccel/ast/c_concepts.py

```diff
@@ -34,3 +35,149 @@ def obj(self):
         """The object whose address is of interest
         """
         return self._obj
+
+#------------------------------------------------------------------------------
+class CStringExpression(Basic):
+    """
+    Internal class used to hold a C string that has LiteralStrings and C macros
+    Parameters:
+        *args: str or LiteralString or CMacro or CStringExpression
+            any number of arguments to be added to the expression
+            note: they will get added in the order provided
+    Example:
+        CStringExpression(
+            CMacro("m"),
+            CStringExpression(
+                LiteralString("the macro is: "),
+                CMacro("mc")
+            ),
+            LiteralString("."),
+        )
+    """
+    __slots__  = ('_expression',)
+    _attribute_nodes  = ()
```

Comment:
I don't exactly know what the `_attribute_nodes` do
is there documentation for it?

Reply:
Unfortunately not yet. It stores all pyccel objects which are contained in the object. It is used in the functions [`get_user_nodes`](https://github.com/pyccel/pyccel/blob/master/pyccel/ast/basic.py#L92-L107), [`get_attribute_nodes`](https://github.com/pyccel/pyccel/blob/master/pyccel/ast/basic.py#L122-L137), [`is_attribute_of`](https://github.com/pyccel/pyccel/blob/master/pyccel/ast/basic.py#L169-L181), [`is_user_of`](https://github.com/pyccel/pyccel/blob/master/pyccel/ast/basic.py#L184-L198) and [`substitute`](https://github.com/pyccel/pyccel/blob/master/pyccel/ast/basic.py#L221-L237) which all have docstrings.

Reply:
from what I can see the _attribute_nodes make the `super.__init__()` [converts any iterable ](https://github.com/pyccel/pyccel/blob/3dcd437333ea618ff2081f5c9d437113078f63b1/pyccel/ast/basic.py#L51)(the list `_expression`) to a tuple, so the `.append` is not working anymore, do you thing it's a good idea to use something like: 
```python
self._expression += (o,)
```

Reply:
For example you could use the following to determine if `#include <inttypes.h>` is needed:
```
int64_macro = CMacro("PRId64")
...
if expr.is_user_of(int64_macro):
```
but if `_attribute_nodes` is empty then it won't recognise that `expr` uses `int64_macro`

Reply:
> from what I can see the _attribute_nodes make the `super.__init__()` [converts any iterable ](https://github.com/pyccel/pyccel/blob/3dcd437333ea618ff2081f5c9d437113078f63b1/pyccel/ast/basic.py#L51)(the list `_expression`) to a tuple, so the `.append` is not working anymore, do you thing it's a good idea to use something like:
> 
> ```python
> self._expression += (o,)
> ```

I tried this and it works fine but is not sure if this is how it should be done. what do you think?

Reply:
This is done for 2 reasons. Firstly to ensure that objects are compatible rather than sometimes being lists and sometimes being tuples, and secondly to remind the developer to take care of the inheritance.
When you do:
```
self._expression += (o,)
```
you should also do:
```
o.set_current_user_node(self)
```

Reply:
> > from what I can see the _attribute_nodes make the `super.__init__()` [converts any iterable ](https://github.com/pyccel/pyccel/blob/3dcd437333ea618ff2081f5c9d437113078f63b1/pyccel/ast/basic.py#L51)(the list `_expression`) to a tuple, so the `.append` is not working anymore, do you thing it's a good idea to use something like:
> > ```python
> > self._expression += (o,)
> > ```
> 
> I tried this and it works fine but is not sure if this is how it should be done. what do you think?

I think this is related to [this discussion](https://github.com/pyccel/pyccel/pull/1181#discussion_r970025146)